extensions/product-recs/bundle/src/main/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServlet.java

```diff
@@ -0,0 +1,178 @@
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.commerce.extensions.recommendations.internal.servlets;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+import javax.servlet.Servlet;
+import javax.servlet.ServletException;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.SlingHttpServletResponse;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.SyntheticResource;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.api.servlets.SlingSafeMethodsServlet;
+import org.apache.sling.api.wrappers.ValueMapDecorator;
+import org.osgi.service.component.annotations.Component;
+
+import com.adobe.granite.ui.components.Value;
+import com.adobe.granite.ui.components.ds.DataSource;
+import com.adobe.granite.ui.components.ds.SimpleDataSource;
+import com.day.cq.wcm.api.policies.ContentPolicy;
+import com.day.cq.wcm.api.policies.ContentPolicyManager;
+import com.drew.lang.annotations.NotNull;
+
+@Component(
+    service = { Servlet.class },
+    property = {
+        "sling.servlet.resourceTypes=" + AllowedRecTypesDataSourceServlet.RESOURCE_TYPE_V1,
+        "sling.servlet.methods=GET",
+        "sling.servlet.extensions=html"
+    })
+public class AllowedRecTypesDataSourceServlet extends SlingSafeMethodsServlet {
+
+    public final static String RESOURCE_TYPE_V1 = "core/cif/extensions/product-recs/datasources/allowedrectypes/v1";
+    public final static String PN_ALLOWED_TYPES = "allowedRecTypes";
+
+    @Override
+    protected void doGet(@NotNull SlingHttpServletRequest request, @NotNull SlingHttpServletResponse response)
+        throws ServletException, IOException {
+        SimpleDataSource allowedRecTypesDataSource = new SimpleDataSource(getAllowedRecTypes(request).iterator());
+        request.setAttribute(DataSource.class.getName(), allowedRecTypesDataSource);
+    }
+
+    /**
+     * Returns a list of allowed product recommendation types based on the template policy.
+     */
+    private List<Resource> getAllowedRecTypes(@NotNull SlingHttpServletRequest request) {
+        List<Resource> allowedRecTypes = new ArrayList<>();
+
+        ResourceResolver resolver = request.getResourceResolver();
+        Resource contentResource = resolver.getResource((String) request.getAttribute(Value.CONTENTPATH_ATTRIBUTE));
+        ContentPolicyManager policyManager = resolver.adaptTo(ContentPolicyManager.class);
+        if (contentResource != null && policyManager != null) {
```

Comment:
You may want to use Stream/Optional APIs here to keep the nested depth low(er)

Reply:
Done.